build.zig

```diff
@@ -2,42 +2,119 @@
+    buildDebug(b);
+}
 
-    const run_qemu_debug = b.addSystemCommand([][]const u8 {
-        "qemu-system-i386",
-        "-display", "curses",
-        "-kernel", kernel.getOutputPath(),
-        "-s", "-S",
+// gdb -ex "symbol-file bin/iso/boot/pluto.elf" -ex "target remote localhost:1234"
+fn buildDebug(b: *Builder) void {
+    const step = b.step("debug", "Debug with gdb");
+    const cmd = b.addSystemCommand([][]const u8{
+        "gdb",
+        "-ex",
+        "symbol-file bin/iso/boot/pluto.elf",
+        "-ex",
+        "target remote localhost:1234",
     });
+    step.dependOn(&cmd.step);
+}
+
+fn buildRun(b: *Builder, target: builtin.Arch, build_path: []const u8, iso_path: []const u8) void {
+    const step = b.step("run", "Run with qemu");
+    const qemu = if (target == builtin.Arch.i386) "qemu-system-i386" else unreachable;
+    var qemu_flags = ArrayList([]const u8).init(b.allocator);
+    qemu_flags.appendSlice([][]const u8{
+        qemu,
+        "-cdrom",
+        iso_path,
+        "-boot",
+        "d",
+        "-serial",
+        "stdio",
+    }) catch unreachable;
+    if (debug)
+        qemu_flags.appendSlice([][]const u8{
+            "-s",
+            "-S",
+        }) catch unreachable;
+    const cmd = b.addSystemCommand(qemu_flags.toSlice());
+    step.dependOn(&cmd.step);
+}
+
+fn buildISO(b: *Builder, build_path: []const u8, iso_path: []const u8) ArrayList(*Step) {
+    const grub_build_path = concat(arena, build_path, "/iso/boot/") catch unreachable;
+    const iso_dir_path = concat(arena, build_path, "/iso") catch unreachable;
+    const grub_cmd = b.addSystemCommand([][]const u8{ "cp", "-r", "grub", grub_build_path.toSlice() });
+    const iso_cmd = b.addSystemCommand([][]const u8{ "grub-mkrescue", "-o", iso_path, iso_dir_path.toSlice() });
+    var steps = ArrayList(*Step).init(b.allocator);
+    steps.append(&grub_cmd.step) catch unreachable;
+    steps.append(&iso_cmd.step) catch unreachable;
+    return steps;
+}
 
-    run_qemu.step.dependOn(&kernel.step);
-    //run_qemu_debug.step.dependOn(&kernel.step);
+fn buildLink(b: *Builder, target: builtin.Arch, build_path: []const u8) *Step {
+    const exec = b.addExecutable("pluto.elf", null);
+    const elf_path = concat(arena, build_path, "/iso/boot") catch unreachable;
+    exec.setOutputDir(elf_path.toSlice());
+    exec.setLinkerScriptPath("link.ld");
+    exec.setTarget(target, builtin.Os.freestanding, builtin.Abi.gnu);
+    for (src_files) |file| {
+        var file_obj = concat(arena, build_path, "/") catch unreachable;
+        file_obj.append(file) catch unreachable;
+        file_obj.append(".o") catch unreachable;
+        exec.addObjectFile(file_obj.toSlice());
+    }
+    return &exec.step;
+}
 
-    b.default_step.dependOn(&run_qemu.step);
-}
\ No newline at end of file
+fn buildObjects(b: *Builder, target: builtin.Arch, build_path: []const u8, src_path: []const u8) ArrayList(*Step) {
```

Comment:
But we'll have to build some files individually, such as architecture specific code and the boot code as otherwise it will be removed due to lazy loading.